In [1]:
#Actividad final modulo 7

#### 1.Cargue las librerías que necesita

In [1]:
# Tratamiento de datos
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Input, Dense
from keras.utils import to_categorical
from keras.models import Model
from sklearn.preprocessing import StandardScaler

#### 2. Importe la base de datos pima-indians-diabetes.csv

In [3]:
df = pd.read_csv('pima-indians-diabetes.csv', header= None)

df.columns = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']

df.head(5)

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1
3     1    89    66    23    94  28.1  0.167   21      0
4     0   137    40    35   168  43.1  2.288   33      1

#### 3. Defina la Base de datos, incluyendo medidas descriptivas y las variables que posee (puede usar las data .names). Considere que la variable de salida (y) es 0,1.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   preg    768 non-null    int64  
 1   plas    768 non-null    int64  
 2   pres    768 non-null    int64  
 3   skin    768 non-null    int64  
 4   test    768 non-null    int64  
 5   mass    768 non-null    float64
 6   pedi    768 non-null    float64
 7   age     768 non-null    int64  
 8   class   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [6]:
df.describe()

preg        plas        pres        skin        test        mass  \
count  768.000000  768.000000  768.000000  768.000000  768.000000  768.000000   
mean     3.845052  120.894531   69.105469   20.536458   79.799479   31.992578   
std      3.369578   31.972618   19.355807   15.952218  115.244002    7.884160   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      1.000000   99.000000   62.000000    0.000000    0.000000   27.300000   
50%      3.000000  117.000000   72.000000   23.000000   30.500000   32.000000   
75%      6.000000  140.250000   80.000000   32.000000  127.250000   36.600000   
max     17.000000  199.000000  122.000000   99.000000  846.000000   67.100000   

             pedi         age       class  
count  768.000000  768.000000  768.000000  
mean     0.471876   33.240885    0.348958  
std      0.331329   11.760232    0.476951  
min      0.078000   21.000000    0.000000  
25%      0.243750   24.000000    0.000000  
50%      0.372500   29.000000    0.000000  
75%      0.626250   41.000000    1.000000  
max      2.420000   81.000000    1.000000

#### 4. Defina el modelo Keras

In [7]:
X = df.drop('class', axis=1)
y = df['class']

y = to_categorical(y)

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [9]:
scaler = StandardScaler()
scaler.fit(X_train)

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

#### 5. Utilice la función de activación de la unidad lineal rectificada denominada ReLU en las dos primeras capas y la función Sigmoid en la capa de salida.

In [10]:
model = Sequential()
model.add(Dense(16, input_dim=X.shape[1], activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(2, activation='sigmoid'))

#### 6. Compile el modelo Keras

In [11]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [12]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 16)                144       
                                                                 
 dense_1 (Dense)             (None, 8)                 136       
                                                                 
 dense_2 (Dense)             (None, 2)                 18        
                                                                 
Total params: 298
Trainable params: 298
Non-trainable params: 0
_________________________________________________________________


#### 7. Ajuste el modelo, considere epochs=150, batch_size=10

In [13]:
model.fit(X_train_scaled, y_train, epochs=150, batch_size=10, verbose=1)

Epoch 1/150
62/62 [==============================] - 3s 4ms/step - loss: 0.7040 - accuracy: 0.4951
Epoch 2/150
62/62 [==============================] - 0s 2ms/step - loss: 0.6389 - accuracy: 0.6401
Epoch 3/150
62/62 [==============================] - 0s 2ms/step - loss: 0.5903 - accuracy: 0.6596
Epoch 4/150
62/62 [==============================] - 0s 2ms/step - loss: 0.5553 - accuracy: 0.7052
Epoch 5/150
62/62 [==============================] - 0s 3ms/step - loss: 0.5299 - accuracy: 0.7296
Epoch 6/150
62/62 [==============================] - 0s 3ms/step - loss: 0.5117 - accuracy: 0.7492
Epoch 7/150
62/62 [==============================] - 0s 4ms/step - loss: 0.4982 - accuracy: 0.7492
Epoch 8/150
62/62 [==============================] - 0s 4ms/step - loss: 0.4885 - accuracy: 0.7541
Epoch 9/150
62/62 [==============================] - 0s 4ms/step - loss: 0.4799 - accuracy: 0.7622
Epoch 10/150
62/62 [==============================] - 0s 3ms/step - loss: 0.4742 - accuracy: 0.7752
Epoch 11/

#### 8. Evalúe el accuracy de su modelo, comente.

In [14]:
test_loss, test_acc = model.evaluate(X_test_scaled, y_test)
print('Test accuracy:', test_acc)

5/5 [==============================] - 0s 2ms/step - loss: 0.7187 - accuracy: 0.7208
Test accuracy: 0.7207792401313782


El accuracy es bajo aunque puede deberse a factores notables como la cantidad de registros que es muy baja (768), para determinar si este modelo es aceptable o no habria que compararlo con otro cambiando los valores de hiperparametros, filtrando parametros y usando tecnicas de ensamble ademas de probar un modelo con mayor cantidad de registros.